btree/btree.c

```diff
@@ -3529,164 +3565,561 @@ int64_t btree_iter_tick(struct m0_sm_op *smop)
+ * @return int8_t return -1 if any ancestor node is not valid. return 1, if
+ *                loading of child is needed, else return 0;
  */
-void m0_btree_iter(struct m0_btree *arbor, const struct m0_btree_key *key,
-		   const struct m0_btree_cb *cb, uint64_t flags,
-		   struct m0_btree_op *bop)
-{
-	m0_sm_op_init(&bop->bo_op, &btree_iter_tick, &bop->bo_op_exec,
-		      &btree_conf, &bop->bo_sm_group);
+static int8_t root_child_is_req(struct m0_btree_op *bop)
+{
+	struct m0_btree_oimpl *oi = bop->bo_i;
+	int8_t                 load = 0;
+	int                    used_count = oi->i_used;
+	do {
+		if (!node_is_valid(oi->i_level[used_count].l_node))
+			return -1;
+		if (used_count == 0) {
+			load = 1;
+			break;
+		}
+		if (!node_isunderflow(oi->i_level[used_count].l_node, true))
+			break;
+
+		used_count--;
+	}while (1);
+	return load;
 }
 
 /**
- * m0_btree_put is the API which is used by motr to put the given record into
- * the btree.
+ * This function will get called if root is an internal node and it contains
+ * only two records. It will check if there is requirement for loading root's
+ * other child and accordingly return the next state for execution.
  *
- * @param arbor It provides all the required data about the tree in which user
- * wants to insert record.
- * @param tx It represents the transaction of which the current operation is
- * part of.
- * @param rec It represents the record which needs to get inserted. Note that,
- * user may or may not provide valid value but record should be provided with
- * valid key, key size and value size as it is needed for operation.
- * @param cb It represents callback which will get called by the operation to
- * allow user to insert key and record at provided place.
- * @param flags
- * @param bop It represents the structure containing all the relevant details
- * for carrying out btree operations.
+ * @param bop will provide all required information about btree operation.
+ * @return int64_t return state which needs to get executed next.
  */
-void m0_btree_put(struct m0_btree *arbor, struct m0_be_tx *tx,
-		  const struct m0_btree_rec *rec,
-		  const struct m0_btree_cb *cb, uint64_t flags,
-		  struct m0_btree_op *bop)
+static int64_t root_case_handle(struct m0_btree_op *bop)
 {
-	bop->bo_opc    = M0_BO_PUT;
-	bop->bo_arbor  = arbor;
-	bop->bo_rec    = *rec;
-	bop->bo_cb     = *cb;
-	bop->bo_tx     = tx;
-	bop->bo_flags  = flags;
-	bop->bo_i      = NULL;
+	/**
+	 * If root is an internal node and it contains only two records, check
+	 * if any record is going to be deleted if yes, we also have to load
+	 * other child of root so that we can copy the content from that child
+	 * at root and decrease the level by one.
+	 */
+	struct m0_btree_oimpl *oi = bop->bo_i;
+	int8_t                 load;
 
-	m0_sm_op_init(&bop->bo_op, &btree_put_tick, &bop->bo_op_exec,
-		   &btree_conf, &bop->bo_sm_group);
-}
+	load = root_child_is_req(bop);
+	if (load == -1) {
+		level_cleanup(oi, bop->bo_tx);
```

Comment:
We can use m0_sm_op_sub but there will be overhead of execution of extra phase i.e. P_CLEANUP. 
and here we will depend on sm implementation .when there will be btree operations by multiple thread, sm may put this as a pending operation if there is other process which has more priority. So, here I was thinking to avoid execution of extra phases if possible